src/blog/understanding-ecmascript-part-4.md

```diff
@@ -0,0 +1,280 @@
+---
+title: 'Understanding the ECMAScript spec, part 4'
+author: '[Marja Hölttä](https://twitter.com/marjakh), speculative specification spectator'
+avatars:
+  - marja-holtta
+date: 2020-05-06
+tags:
+  - ECMAScript
+description: 'Tutorial on reading the ECMAScript specification'
+tweet: ''
+---
+
+## Previous episodes
+
+In [part 1](/blog/understanding-ecmascript-part-1), we read through a simple method — `Object.prototype.hasOwnProperty` — and the **abstract operations** it invokes. We familiarized ourselves with the shorthands `?` and `!` related to error handling. We encountered **language types**, **specification types**, **internal slots**, and **internal methods**.
+
+In [part 2](/blog/understanding-ecmascript-part-2), we examined a simple grammar production and how its runtime semantics are defined. In [the extra content](/blog/extra/understanding-ecmascript-part-2-extra), we also followed a long grammar production chain from `AssignmentExpression` to `MemberExpression`.
+
+In [part 3](/blog/understanding-ecmascript-part-3), we familiarized ourselves with the lexical grammar, the syntactic grammar, and the shorthands used for defining the syntactic grammar.
+
+## Meanwhile in other parts of the Web
+
+[Jason Orendorff](https://github.com/jorendorff) from Mozilla published [a great in-depth analysis of JS syntactic quirks](https://github.com/mozilla-spidermonkey/jsparagus/blob/master/js-quirks.md#readme). Even though the implementation details differ, every JS engine faces the same problems with these quirks.
+
+## Cover grammars
+
+In this episode, we take a deeper look into *cover grammars*. They are a way to specify grammar rules for syntactic constructs where we don't know what we're looking at until we've seen the complete construct.
+
+Again, we'll skip the subscripts for `[In, Yield, Await]` for brevity, as they aren't important for this blog post. See [part 3](/blog/understanding-ecmascript-part-3) for an explanation of their meaning and usage.
+
+## Parenthesized expression or an arrow parameter list?
+
+Typically, parsers decide which grammar production to follow based on finite lookahead.
```

Comment:
I think it would be useful to show a quick example of simple finite lookahead first (maybe something like the start of an `ExpressionStatement`?), and go into a complex one (that requires cover grammar) after.

Reply:
You mean the negative lookahead? I wouldn't want to confuse this blog post with negative lookaheads, as that's not related to cover grammars.

Reply:
... if I'd like to have an example of a finite positive lookahead, UnaryExpression would be it. But I'm not sure if explaining that brings anything at this point. A more interesting concept would be the "we're not yet know which production to take, but we know the next symbol, so let's parse the next symbol and decide after that" which occurs e.g., for CallExpression, Expression, etc. Not sure if adding that would be beneficial either.

Reply:
Positive lookahead sounds good too; I just think it's useful to show a simple lookahead example to go with that sentence.

Reply:
Added lookahead examples.